llama.cpp

```diff
@@ -11684,6 +11767,214 @@ struct llm_build_context {
+                    Qcur = ggml_add(ctx0, Qcur, model.layers[il].bq);
+                    cb(Qcur, "Qcur", il);
+                }
+
+                // B1.K
+                struct ggml_tensor * Kcur = ggml_mul_mat(ctx0, model.layers[il].wk, cur);
+                Kcur = ggml_mul(ctx0, Kcur, model.layers[il].wk_scale);
+                cb(Kcur, "Kcur", il);
+                if (model.layers[il].bk) {
+                    Kcur = ggml_add(ctx0, Kcur, model.layers[il].bk);
+                    cb(Kcur, "Kcur", il);
+                }
+
+                // B1.V
+                struct ggml_tensor * Vcur = ggml_mul_mat(ctx0, model.layers[il].wv, cur);
+                Vcur = ggml_mul(ctx0, Vcur, model.layers[il].wv_scale);
+                cb(Vcur, "Vcur", il);
+                if (model.layers[il].bv) {
+                    Vcur = ggml_add(ctx0, Vcur, model.layers[il].bv);
+                    cb(Vcur, "Vcur", il);
+                }
+
+                Qcur = ggml_rope_ext(
+                    ctx0, ggml_reshape_3d(ctx0, Qcur, n_embd_head, n_head, n_tokens), inp_pos, nullptr,
+                    n_rot, rope_type, n_ctx_orig, freq_base, freq_scale,
+                    ext_factor, attn_factor, beta_fast, beta_slow
+                );
+                cb(Qcur, "Qcur", il);
+
+                Kcur = ggml_rope_ext(
+                    ctx0, ggml_reshape_3d(ctx0, Kcur, n_embd_head, n_head_kv, n_tokens), inp_pos, nullptr,
+                    n_rot, rope_type, n_ctx_orig, freq_base, freq_scale,
+                    ext_factor, attn_factor, beta_fast, beta_slow
+                );
+                cb(Kcur, "Kcur", il);
+
+                llm_build_kv_store(ctx0, hparams, cparams, kv_self, gf, Kcur, Vcur, n_tokens, kv_head, cb, il);
+
+                const int64_t n_ctx                 = cparams.n_ctx;
+                const int64_t n_head                = hparams.n_head;
+                const int64_t n_head_kv             = hparams.n_head_kv;
+                const int64_t n_embd_head_k         = hparams.n_embd_head_k;
+                const int64_t n_embd_k_gqa          = hparams.n_embd_k_gqa();
+                const int64_t n_embd_head_v         = hparams.n_embd_head_v;
+                const int64_t n_embd_v_gqa          = hparams.n_embd_v_gqa();
+
+                float                      kq_scale = 1.0f/sqrtf(float(n_embd_head));
+                struct ggml_tensor *       cur_attn;
+                struct ggml_tensor *              q = ggml_permute(ctx0, Qcur, 0, 2, 1, 3);
+                cb(q, "q", il);
+
+                struct ggml_tensor * k =
+                    ggml_view_3d(ctx0, kv_self.k_l[il],
+                            n_embd_head_k, n_kv, n_head_kv,
+                            ggml_row_size(kv_self.k_l[il]->type, n_embd_k_gqa),
+                            ggml_row_size(kv_self.k_l[il]->type, n_embd_head_k),
+                            0);
+                cb(k, "k", il);
+
+                if (cparams.flash_attn) {
```

Comment:
Is there any reason to duplicate all this code here instead of calling `llm_build_kqv`?

Reply:
Bitnet Attention structure is a little different from normal attention, that there is a `attn_sub_norm` in BitNet attention, before `wo mul_mat`. `llm_build_kqv` only suits for normal attention, and does not have a param for `attn_sub_norm`, so unless we change the l`lm_build_kqv` api, seems the only choice is to duplicate the code to add the `attn_sub_norm` in `llm_build_kqv` (the same situation as `llm_build_ffn`), is there a better way for reuse the `llm_build_kqv`?

Reply:
As far as I can tell, the only difference is the `ggml_mul` with `wo_scale` at the end. You could do this after `llm_build_kqv`, or add a `wo_scale` parameter to the function.

Reply:
Am I understanding it correct that I need to add `attn_sub_norm` and `wo_scale` parameters to `llm_build_kv ` function,  `ffn_sub_norm` / `ffn_gate_scale` / `ffn_up_scale` / `ffn_down_scale` to `llm_build_ffn` function to suits BitNet, and use `llm_build_kv`, `llm_build_ffn` to build the BitNet? Considered it may change the code for all models use these two functions, I want to double check it.

Reply:
A less intrusive way may be to modify `llm_build_kqv` to skip the `wo` mul_mat if it is `NULL`, and do the `attn_sub_norm` and `wo` steps these after the call to `llm_build_kv` in `llm_build_bitnet`. Let's leave `llm_build_ffn` as is for now.

Reply:
Ok. I'll modify `llm_build_kqv` by adding a `if (wo)` and passing a nullptr as `wo` in BitNet to let it computed outside of the `llm_build_kv`, and leave the ffn code without changing.